src/qibo/base/hamiltonians.py

```diff
@@ -187,6 +211,179 @@ def __matmul__(self, o):
                                              "implemented.".format(type(o)))
 
 
+class SymbolicHamiltonian:
+    """Parses symbolic Hamiltonians defined using ``sympy``.
+
+    This class should not be used by users.
+    It is used internally to help creating
+    :class:`qibo.base.hamiltonians.Hamiltonian` and
+    :class:`qibo.base.hamiltonians.TrotterHamiltonian` objects for Hamiltonians
+    defined using symbols. For more information we refer to the
+    :meth:`qibo.base.hamiltonians.Hamiltonian.from_symbolic`
+    and :meth:`qibo.base.hamiltonians.TrotterHamiltonian.from_symbolic` methods.
+
+    Args:
+        symbolic_hamiltonian: The full Hamiltonian written with symbols.
+        symbol_map (dict): Dictionary that maps each symbol to a pair of
+            (target, matrix).
+    """
+    import sympy
+    from qibo import matrices
+
+    def __init__(self, hamiltonian, symbol_map):
+        self.symbolic = self.sympy.expand(hamiltonian)
+        self.map = symbol_map
+
+        self.constant = 0
+        self.terms = dict()
+        target_ids = set()
+        for term in self.symbolic.args:
+            if term.args:
+                expression = term.args
+            else:
+                expression = (term,)
+
+            symbols = [x for x in expression if x.is_symbol]
+            numbers = [x for x in expression if not x.is_symbol]
+            if numbers:
+                assert len(numbers) == 1
+                const = float(numbers[0])
```

Comment:
Is float save or shall we prefer DTYPE (numpy)?

Reply:
It was working with `float` but I changed that to `self.matrices.dtype` since this number is multiplied/added to matrices. I am not really sure which choice is generally safer/better.